sw/device/lib/runtime/otbn.c

```diff
@@ -17,14 +17,29 @@ const int kOtbnWlenBytes = 256 / 8;
 otbn_result_t otbn_data_ptr_to_dmem_addr(const otbn_t *ctx, otbn_ptr_t ptr,
                                          uint32_t *dmem_addr_otbn) {
   uintptr_t ptr_addr = (uintptr_t)ptr;
-  uintptr_t app_dmem_start_addr = (uintptr_t)ctx->app.dmem_start;
-  uintptr_t app_dmem_end_addr = (uintptr_t)ctx->app.dmem_end;
+  uintptr_t app_dmem_data_start_addr = (uintptr_t)ctx->app.dmem_data_start;
+  uintptr_t app_dmem_data_end_addr = (uintptr_t)ctx->app.dmem_data_end;
+  uintptr_t app_dmem_bss_start_addr = (uintptr_t)ctx->app.dmem_bss_start;
+  uintptr_t app_dmem_bss_end_addr = (uintptr_t)ctx->app.dmem_bss_end;
+  uintptr_t app_dmem_bss_offset = (uintptr_t)ctx->app.dmem_bss_offset;
 
-  if (dmem_addr_otbn == NULL || ptr == NULL || ctx == NULL ||
-      ptr_addr < app_dmem_start_addr || ptr_addr > app_dmem_end_addr) {
+  if (dmem_addr_otbn == NULL || ptr == NULL || ctx == NULL) {
     return kOtbnBadArg;
   }
-  *dmem_addr_otbn = ptr_addr - app_dmem_start_addr;
+
+  if (app_dmem_data_start_addr <= ptr_addr &&
+      ptr_addr < app_dmem_data_end_addr) {
+    // Pointer is in the data section, which is at the start of DMEM
+    *dmem_addr_otbn = ptr_addr - app_dmem_data_start_addr;
+  } else if (app_dmem_bss_start_addr <= ptr_addr &&
+             ptr_addr < app_dmem_bss_end_addr) {
+    // Pointer is in the bss section, which is after the data section
+    *dmem_addr_otbn = ptr_addr - app_dmem_bss_start_addr + app_dmem_bss_offset;
+  } else {
+    // Pointer is not in a valid DMEM region
+    return kOtbnBadArg;
+  }
```

Comment:
This could be dramatically simplified if you could just assume that OTBN .bss came just after OTBN .data in the Ibex link. Is that not true?